File: NEsper/events/property/ArrayMethodPropertyGetter.cs
using System;
using System.Reflection;

using net.esper.events;

namespace net.esper.events.property
{
	/// <summary>
	/// Getter for an array property identified by a given index, using vanilla reflection.
	/// </summary>
	
	public class ArrayMethodPropertyGetter : EventPropertyGetter
	{
		private readonly MethodInfo method;
		private readonly int index;

		/// <summary>
		/// Constructor.
		/// </summary>
		/// <param name="method">is the method to use to retrieve a value from the object</param>
		/// <param name="index">is tge index within the array to get the property from</param>
		
		public ArrayMethodPropertyGetter(MethodInfo method, int index)
		{
			this.index = index;
			this.method = method;
			
			if (index < 0)
			{
				throw new ArgumentException("Invalid negative index value");
			}
		}

		/// <summary>
		/// Gets the value.
		/// </summary>
		/// <param name="obj">The obj.</param>
		/// <returns></returns>
		
		public Object GetValue(EventBean obj)
		{
			Object underlying = obj.Underlying;
			
			try
			{
				Object value = method.Invoke(underlying, (Object[]) null);
				if (((Array) value).Length <= index)
				{
					return null;
				}

				return ((Array) value).GetValue(index);
			}
			catch (InvalidCastException)
			{
				throw new PropertyAccessException("Mismatched getter instance to event bean type");
			}
			catch (TargetInvocationException e)
			{
				throw new PropertyAccessException(e);
			}
			catch (UnauthorizedAccessException e)
			{
				throw new PropertyAccessException(e);
			}
			catch (ArgumentException e)
			{
				throw new PropertyAccessException(e);
			}
		}
		
		public override String ToString()
		{
			return "ArrayMethodPropertyGetter " + " method=" + method.ToString() + " index=" + index;
		}
	}
}

File: trunk/esper/src/main/java/com/espertech/esper/event/property/ArrayMethodPropertyGetter.java
package com.espertech.esper.event.property;

import com.espertech.esper.event.EventBean;
import com.espertech.esper.event.EventPropertyGetter;
import com.espertech.esper.event.PropertyAccessException;

import java.lang.reflect.Array;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * Getter for an array property identified by a given index, using vanilla reflection.
 */
public class ArrayMethodPropertyGetter implements EventPropertyGetter
{
    private final Method method;
    private final int index;

    /**
     * Constructor.
     * @param method is the method to use to retrieve a value from the object
     * @param index is tge index within the array to get the property from
     */
    public ArrayMethodPropertyGetter(Method method, int index)
    {
        this.index = index;
        this.method = method;

        if (index < 0)
        {
            throw new IllegalArgumentException("Invalid negative index value");
        }
    }

    public final Object get(EventBean obj) throws PropertyAccessException
    {
        Object underlying = obj.getUnderlying();

        try
        {
            Object value = method.invoke(underlying, null);
            if (Array.getLength(value) <= index)
            {
                return null;
            }
            return Array.get(value, index);
        }
        catch (ClassCastException e)
        {
            throw new PropertyAccessException("Mismatched getter instance to event bean type");
        }
        catch (InvocationTargetException e)
        {
            throw new PropertyAccessException(e);
        }
        catch (IllegalAccessException e)
        {
            throw new PropertyAccessException(e);
        }
        catch (IllegalArgumentException e)
        {
            throw new PropertyAccessException(e);
        }
    }

    public String toString()
    {
        return "ArrayMethodPropertyGetter " +
                " method=" + method.toString() +
                " index=" + index;
    }

    public boolean isExistsProperty(EventBean eventBean)
    {
        return true; // Property exists as the property is not dynamic (unchecked)
    }    
}
